Ported to 7.3

IsLocal opened the pull request:
### Fully Ported ✅
I haven't been able to bug test on any other platforms but, using 2 instances of BTA 7.3_04 on my pc running windows 11 it works without error (I can hear my self).


  I ported the project environment to 7.3, fixed up the code so it doesn't crash on 7.3 apis, and added some client sided features like the volume mixer and mic indicator.

This is my first time coding in java and my first time contributing on here so I think it **reaaallly** does need some feedback.

IsLocal committed:
(hackely) updated code to atleast load into 7.3_01.

IsLocal committed:
Updated for 7.3 and added talking state just because

IsLocal committed:
Added Volume Mixer, volume mixed saving to config, fixed some crashes, added mic indicator, cleaned up logs, reintroduced crash fix.

5Avian commented:
Ooh sick, thanks for the PR! I'll debug my local environment to see why in the world it's failing, and I'll get back to you ASAP.

5Avian commented:
A lot of code has been changed I see. Before going into implementation specifics, there's a couple pieces of feedback i wanna give:

- There's a bunch of minor changes to formatting, think if-statements written in one line instead of two, and added whitespace. This mod is my personal project, so it's written in my code style. It always goes appreciated when you try to keep your code in a similar style to mine, and update as little of the existing code as necessary.
- There's no summary of what you did in the PR, which could help me a lot as the reviewer to understand exactly what you did. The commit naming is helpful, but I usually don't know enough about a feature only from it's name.
- You've added a lot of functionality in the same PR as a port. Especially when porting a version (which already has a bunch of things that can go wrong), it's quite risky to add a bunch of new features at the same time. I haven't experienced the BTA 7.3 codebase before, so on top of understanding the changes that they made for this version, I also need to understand how it's being used within your features.

I'll review each file and feature later. Thanks for the contribution, and I hope my feedback will help you. If I were to merge this PR, I'd like to do it step-by-step. First the port, then once I've tested that it fully works, the QOL features.

5Avian commented on build.gradle:
```diff
@@ -1,75 +1,126 @@
+		metadataSources { artifact() }
 	}
 	ivy {
-		name = "Fabric Loader, HalpLibe, Mod Menu"
```
Keeping these names in would be appreciated, because they help me understand which dependencies are in which repositories.

5Avian commented on src/main/resources/fabric.mod.json:
```diff
@@ -1,7 +1,7 @@
 {
   "schemaVersion": 1,
   "id": "proxvc",
-  "version": "${mod_version}",
```
This was done automatically before. I'm assuming here, but I think this broke during your development because you changed a line at the bottom of build.gradle. Please undo this. I'll leave a comment in build.gradle showing the issue.

5Avian commented on src/main/java/fiveavian/proxvc/mixin/client/ALC10Mixin.java:
What is your reasoning behind removing this? Are you 100% sure that this bug is completely fixed in newer versions? I spent a bunch of time scratching my head coming up with this hacky solution. It only appears on some devices, and I don't remember what the reason was. Maybe foreign languages using Chinese (or similar non-Latin) character sets, but don't quote me on that. If you're unsure whether this is now fixed, I'd prefer to keep this in.

5Avian commented on src/main/java/fiveavian/proxvc/mixin/client/NetClientHandlerMixin.java:
```diff
@@ -15,7 +15,7 @@
 
-@Mixin(value = NetClientHandler.class, remap = false)
+@Mixin(value = PacketHandlerClient.class, remap = false)
 public class NetClientHandlerMixin {
```
The reason this class is named NetClientHandlerMixin is because it was a mixin on the NetClientHandler class. Now that it affects the PacketHandlerClient class, could you rename it?

5Avian commented on src/main/java/fiveavian/proxvc/mixin/client/SoundEngineMixer.java:
```diff
@@ -10,10 +10,10 @@
-public class SoundManagerMixin {
-    @Redirect(method = "loadSoundSettings", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/sound/SoundCategoryHelper;isAnyEnabled(Lnet/minecraft/client/option/GameSettings;)Z"))
+@Mixin(value = SoundEngine.class, remap = false)
+public class SoundEngineMixer {
```
This class is a mixin, I'm guessing the rename to Mixer was an oversight. Could you change the name to SoundEngineMixin?

5Avian commented on src/main/java/fiveavian/proxvc/gui/MicrophoneListComponent.java:
```diff
@@ -122,13 +122,16 @@ public void onMouseClick(int mouseButton, int x, int y, int width, int relativeM
 
     @Override
-    public void onMouseMove(int x, int y, int width, int relativeMouseX, int relativeMouseY) {}
+    public void onMouseMove(int x, int y, int width, int relativeMouseX, int relativeMouseY) {
```
I don't feel like collapsing these blocks is necessary. I'd prefer to not have any changes to existing code that is not within the scope of your PR in the future. For now it's fine, you don't have to change it back.

5Avian commented on src/main/java/fiveavian/proxvc/gui/MicrophoneListComponent.java:
```diff
@@ -98,7 +98,7 @@ public void render(int x, int y, int width, int relativeMouseX, int relativeMous
                 GL11.glEnable(GL11.GL_TEXTURE_2D);
             }
-            fontRenderer.drawString(specifier, x + 1, y + 4, 0xFFFFFF);
+            fontRenderer.drawString(specifier.replace("OpenAL Soft on", ""), x + 1, y + 4, 0xFFFFFF);
```
I see 2 possible issues with this code:

- The resulting value of "specifier.replace" does not replace the trailing space character, making the result always start with a space character
- Any microphones called "OpenAL Soft on" in device settings will show up weirdly

They might both be nitpicks, but they do affect the polish. I'd like to atleast see the first issue fixed. If you want you can fix the second one too, but I'd still be willing to merge if the second one stayed in here.

5Avian commented on src/main/java/fiveavian/proxvc/gui/MicrophoneListComponent.java:
```diff
@@ -71,7 +71,7 @@ public int getHeight() {
     public void render(int x, int y, int width, int relativeMouseX, int relativeMouseY) {
         Tessellator tessellator = Tessellator.instance;
-        FontRenderer fontRenderer = mc.fontRenderer;
+        Font fontRenderer = mc.font;
```
The value of the variable no longer abstracts the concept of a renderer, but the concept of a font. Please rename the variable to "font" if possible.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -127,88 +143,138 @@ private void stop(Minecraft client) {
-    private void login(Minecraft client, Packet1Login packet) {
-        Socket socket = client.getSendQueue().netManager.networkSocket;
-        serverAddress = socket.getRemoteSocketAddress();
+    private void login(Minecraft client, PacketLogin packet) {
```
Correct me if I'm wrong, but I assume what you're doing here is recreating AccessWideners with Java's built-in reflection features. This is not something I want to see in the codebase.

Please look into AccessWideners. Basically, they allow you to change scope of private fields without having to write this kind of code. The reason it broke, I assume, is because you seem to have replaced basically the entirety of the build.gradle file with one from a mod template. No judgement, but especially with those things it's important to understand what you're changing. This code's gonna be 10x more understandable once you've done that.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -127,88 +143,138 @@ private void stop(Minecraft client) {
 
     private void disconnect(Minecraft client) {
         serverAddress = null;
+
```
(Some examples of) the whitespace I referred to in my PR comment. Please remove the 3 added empty lines here and the trailing ones in the function above.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -127,88 +143,138 @@ private void stop(Minecraft client) {
-        Socket socket = client.getSendQueue().netManager.networkSocket;
-        serverAddress = socket.getRemoteSocketAddress();
+    private void login(Minecraft client, PacketLogin packet) {
+        // After
```
This comment does not add anything to the codebase, apart from confusion. Please remove it.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -127,88 +143,138 @@ private void stop(Minecraft client) {
             return;
         }
-        GL11.glBindTexture(GL11.GL_TEXTURE_2D, client.renderEngine.getTexture("/proxvc.png"));
+        client.textureManager.loadTexture("/gui/proxvc.png").bind(); // Bind the texture
```
I sadly cannot decompile the code for Minecraft still, so it's hard for me to figure this one out on my own.

Please research whether `loadTexture` loads this texture into memory. If it does, I'd like this to become a field within this class, so we don't waste time loading the same texture 60 times per second. If not, there's no issue here, and this code is fine. Although the comment is arguably useless, since the method call after `loadTexture` is called `bind`.

The reason I'm even commenting this is because `getTexture` implies no work being done except retrieving something, and `loadTexture` implies reading something, from disk or elsewhere.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -127,88 +143,138 @@ private void stop(Minecraft client) {
+            }
             return;
+        }
+
```
CONTRIBUTING.md says: Use no more than 1 consecutive blank line.
Please follow that here.
And please remove some of the blank lines right after the if-statements in the 20 or so lines below this, together with the one roughly 20 lines above this.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -112,8 +128,8 @@ private void start(Minecraft client) {
     private void stop(Minecraft client) {
-        if (optionFilePath != null)
-            OptionStore.saveOptions(optionFilePath, options, keyBindings);
+        if (optionFilePath != null) OptionStore.saveOptions(optionFilePath, options, keyBindings);
```
This line was formatted like that for a reason. It makes lines shorter, making them easier to read on screens, and since it's used everywhere in the code, it's more consistent. If you want to, please fix it.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -90,20 +106,20 @@ private void start(Minecraft client) {
-                    .withComponent(new BooleanOptionComponent(usePushToTalk));
-            OptionsCategory devicesCategory = new OptionsCategory("gui.options.page.proxvc.category.devices")
-                    .withComponent(new MicrophoneListComponent(device, selectedInputDevice));
+            OptionsCategory generalCategory = new OptionsCategory("gui.options.page.proxvc.category.general").withComponent(new FloatOptionComponent(voiceChatVolume)).withComponent(new BooleanOptionComponent(isMuted)).withComponent(new BooleanOptionComponent(usePushToTalk)).withComponent(new BooleanOptionComponent(isDebug));
```
With these lines, I'd prefer them folded out. How it is right now makes it a lot harder to read. Plus, since line 111/112 (before 99/100) were left unchanged, it's now inconsistent.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -90,20 +106,20 @@ private void start(Minecraft client) {
-                    .withComponent(controlsCategory);
+                    .withComponent(new KeyBindingComponent(keyPushToTalk))
+                    .withComponent(new KeyBindingComponent(keyMixer));
+
```
2 consecutive blank lines. Please turn into 1.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -90,20 +106,20 @@ private void start(Minecraft client) {
             device.open(selectedInputDevice.value);
+
+            System.out.println("ProxVC successfully started.");
+
```
Trailing blank line with nothing behind it. This doesn't really add anything to readability. I use blank lines sparingly in my projects, only if there's a very distinct difference in purpose between the code. The indentation below this already makes clear that it has different purpose.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -42,20 +47,24 @@ public class ProxVCClient implements ClientModInitializer {
+    public final KeyBinding[] keyBindings = {keyMute, keyPushToTalk, keyMixer};
+    public OptionFloat voiceChatVolume;
+    public OptionBoolean isMuted;
+    public OptionBoolean isDebug;
```
Does this get used anywhere within code? If not, please remove everything related to this option, including definition. If it does get used, please give it a translation key.

5Avian commented on src/main/java/fiveavian/proxvc/vc/client/VCInputClient.java:
Please remove your dev comments and added blank line in this file.

5Avian commented on build.gradle:
Things to fix (already commented in different files):

- Use access widener
- Make sure version in fabric.mod.json gets inserted automatically (see previous version of this file, at bottom, for example)

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -69,18 +78,25 @@ public void onInitializeClient() {
+
     }
 
+
```
Please remove unnecessary blank lines around here

5Avian commented on src/main/java/fiveavian/proxvc/screens/VolumeMixerScreen.java:
```diff
@@ -0,0 +1,66 @@
+
+        for (StreamingAudioSource source : sources.values()) {
+            if (source.lastHeard < System.currentTimeMillis() - 10000) {
+                continue; // skip sources that haven't been heard from in the last 10 seconds
```
Is this an actually handy thing, to now show sources not heard in the last 10 seconds? If I were to accept this, I'd want this to be configurable or removed.

5Avian commented on src/main/java/fiveavian/proxvc/screens/VolumeMixerScreen.java:
```diff
@@ -0,0 +1,66 @@
+            this.add(mixer);
+            i++;
+        }
+
```
Lots of unnecessary whitespace around here in this file.

5Avian commented on src/main/java/fiveavian/proxvc/screens/VolumeMixerScreen.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    public void keyPressed(char eventCharacter, int eventKey, int mx, int my) {
```
I'm actually not sure, do you need to create this method for the class to be valid? If you can remove this definition, so it uses the `super` method on its own, it'd be much preferred, since there'd be less code, and overriding one of these functions is trivial.Also counts for all the other methods just having a super. call around here.

5Avian commented on src/main/java/fiveavian/proxvc/screens/VolumeMixerScreen.java:
Does this need to be it's own screen? Can't we put this in our options menu? Seems more uniform that way. But the choice is yours.

5Avian commented on src/main/java/fiveavian/proxvc/vc/AudioInputDevice.java:
```diff
@@ -47,19 +60,45 @@ public synchronized ByteBuffer pollSamples() {
+        return isTalking;
+    }
+
+    // did not write this lolll
```
Providing a link to the source of this function would've been handy.

5Avian reviewed (commented):
I've reviewed some of the files changes, but it's a lot for me to do everything at once right now. The most important points I wrote down.

If you wanna merge your code into the project, please do it in 2 steps:

1. PR for port only
2. PR (or issue, so we can discuss implementation) for QoL features

Also, feel free to release your own fork of this mod too. It has the MIT license, so you can basically do anything with it. Of course I'd absolutely prefer to merge port, since this is an existing project, and I'm down to discuss implementation of your QOL features. And don't let my feedback get ya down, the PR is absolutely appreciated!

IsLocal commented:
Thanks for all the feedback and consideration, this is all a learning experience for me so all the feedback only helps. I'll start chipping away at some of the issues mainly just working to cleanup the big stuff e.g. my complete disregard of the build.gradle file and AccessWideners loll. You can tell how little I understood going in to this so like I said feedback only helps.

IsLocal committed:
Resolved issues with build.gradle, reintroduced AccessWideners, fixed socket implementation, reformating, fixed some naming oversights, optimized texture loading.

IsLocal committed:
fixed typo

IsLocal committed:
Prevent trying to read player before loaded.

IsLocal committed:
Removed Mixer Feature from main branch

IsLocal committed:
renamed fontRender variable to font for clarity.

IsLocal committed:
Cleaned up formating for parity with main fork, removed dev comments/logs, and removed redundant functions from port.

Reformatting for parity.
Removed left over mixer lang definition.

IsLocal commented:
Ok so, I cleaned it all up and moved all the other features onto a different branch for a future PR or whatever I end up doing with them and I tried to keep in the scope of just a port for now. The only additional feature not in scope is the isTalking() method along with the icon associated with it.

IsLocal committed:
Reintroduced ALC10Mixin and fixed mixin failure on newer java versions.

IsLocal committed:
Reformating.

IsLocal committed:
fixed incorrect change to loadedEntityList.

IsLocal committed:
corrected redundant build.gradle changes.

IsLocal committed:
reformating and corrected change made in pollSamples().

IsLocal committed:
final touches.

IsLocal commented:
### 7.3 Port <sub>fr this time</sub> 
Heres a summary of the changes I'm now attempting to merge after removing all the random things I added. i def have too much time on my hands having fun doing my first pr.

## Port

- Theres a lot of changed imports, methods, mixins, and fields to interface with the newer 7.3 API correctly.
- Changed directory for status indicator to "resources/gui/proxvc.png" because of what I think is a limitation of 7.3.
    - Changed rendering and loading of texture to use newer Texture class

- Bumped versions in build.gradle for new version along with newer dependencies for lwjgl.
- Bumped versions in gradle.properties.
- Updated dependency repos. 

- Made ALC10Mixin to not be required in runtime to fix a crash I was experiencing related to the mixin.
- Some additional bug fixes made when testing.
  - Making sure the player isn't attempted to be read if not loaded or dead.
   

## Additional implements 

- Changed method to get device specifiers to be more stable using ALUtil class. (which I think fixes the bug with specifiers on older java versions)
- Some changes to OpenAL methods to use newer or more stable ones.
- Additional logging to console for major events.
   - Plus some extra catching on openAL methods to make errors more apparent to the player.

- Automatic specifier reformatting with regex.

- Added talking state to indicator that detects input from mic.
     - IsTalking() method
     - Changes to indicator sprite.
     - Changed uv cords.

5Avian commented on src/main/java/fiveavian/proxvc/ProxVCClient.java:
```diff
@@ -183,25 +189,29 @@ private void render(Minecraft client, WorldRenderer renderer) {
-        Socket socket = client.getSendQueue().netManager.networkSocket;
-        serverAddress = socket.getRemoteSocketAddress();
+    private void login(Minecraft client, PacketLogin packet) {
+            Socket socket = (Socket) client.getSendQueue().netManager.socket;
```
Just see some bad indentation here. These 2 lines should be a tab to the left.

5Avian commented on src/main/java/fiveavian/proxvc/gui/MicrophoneListComponent.java:
```diff
@@ -98,7 +98,7 @@ public void render(int x, int y, int width, int relativeMouseX, int relativeMous
                 GL11.glEnable(GL11.GL_TEXTURE_2D);
             }
-            fontRenderer.drawString(specifier, x + 1, y + 4, 0xFFFFFF);
+            font.drawString(specifier.replaceFirst("OpenAL Soft on\\s+(?!($|\\s+$))", ""), x + 1, y + 4, 0xFFFFFF);
```
Nice that you got this to work! Seems super complex though. I think? regex has a way of only matching the start of the string? Just a thought, could this be simplified to `^OpenAL Soft on `? If not, could you write a short comment above this explaining why it's implemented like this and what it does?
Secondly, IIRC Java compiles the Regex pattern on every render right now. Please check out https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html and see whether you can compile the pattern and store the correct object in the class for later use. It's not really something we *should* be worrying about, because it has minimal performance impact for just one string, but having this stuff consistently using good practices throughout a codebase can help immensely.

5Avian commented on src/main/java/fiveavian/proxvc/mixin/client/ALC10Mixin.java:
```diff
@@ -12,14 +13,15 @@
 /**
  * Ensures that Minecraft doesn't crash on certain devices.
  */
+@Pseudo
```
That's so cool that this exists. Never knew about it. You don't have to do anything with this comment.

5Avian commented on src/main/java/fiveavian/proxvc/vc/StreamingAudioSource.java:
```diff
@@ -10,12 +10,17 @@
     private static final int NUM_BUFFERS = 8;
 
-    public final int source = AL10.alGenSources();
+    public final int source ;
```
Little formatting thing again - there shouldn't be any whitespace between the identifier and semicolon.

5Avian commented on src/main/java/fiveavian/proxvc/vc/StreamingAudioSource.java:
```diff
@@ -10,12 +10,17 @@
+        try {
+            source = AL10.alGenSources();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to generate OpenAL source. Is OpenAL initialized? Check if your volume is set to 0!", e);
```
Is this with computer volume being 0 or Minecraft volume being 0? If it's with Minecraft volume, if this ever happened during testing, there should be an issue with the code in the SoundEngineMixin. The code there should always ensure the game thinks it has its' volume turned on.

IsLocal replied:
Left over from before I added back the mixin I'll just remove it.

5Avian commented on src/main/java/fiveavian/proxvc/vc/client/VCInputClient.java:
```diff
@@ -56,6 +56,9 @@ private void sendNextPacket() throws Exception {
         if (sources.isEmpty()) {
             return;
         }
+        if (client.thePlayer == null) {
```
Could you merge this condition with the one above with an `||`? When the action we take is the same, the line isn't too long and the conditions aren't too different, I usually just add it to existing if-statements to reduce function length.

5Avian reviewed (changes requested):
Awesome! Seems like a solid PR. I pointed out some minor stuff that should probably be fixed to keep everything consistent, and a possible issue with the volume being set to 0. But great stuff, it's really nice to see that people like this enough to work on it themselves. And it's super useful, since I can't do any developing on my own right now.

Also, I tried launching this, but found this error. I assume this has to do with the `build.gradle`? Maybe BTA removed LWJGL from its' binary so that it needs to be added manually? The template project probably has the way to solve it.

```


     Minecraft has crashed!
     ----------------------

Minecraft has stopped running because it encountered a problem.

If you wish to report this, please copy this entire text and post it on the support portal at https://bugs.betterthanadventure.net/
Please include a description of what you did when the error occured.


     System Information
     ------------------

Generated 7/13/25, 11:48 AM
Minecraft: Better than Adventure! 7.3_01
OS: Windows 11 (amd64) version 10.0
Java: 21.0.5, Eclipse Adoptium
VM: OpenJDK 64-Bit Server VM (mixed mode, sharing), Eclipse Adoptium
LWJGL: 3.3.4-snapshot
[failed to get system properties (java.lang.NoClassDefFoundError: Could not initialize class org.lwjgl.system.Library)]


     Error Information
     ------------------

java.lang.UnsatisfiedLinkError: Failed to locate library: lwjgl.dll
	at org.lwjgl.system.Library.loadSystem(Library.java:174)
	at org.lwjgl.system.Library.loadSystem(Library.java:64)
	at org.lwjgl.system.Library.<clinit>(Library.java:52)
	at org.lwjgl.glfw.GLFW.<clinit>(GLFW.java:30)
	at net.minecraft.client.render.window.GameWindowGLFW.init(GameWindowGLFW.java:70)
	at net.minecraft.client.Minecraft.startGame(Minecraft.java:481)
	at net.minecraft.client.Minecraft.run(Minecraft.java:979)
	at net.minecraft.client.Minecraft.startMainThread(Minecraft.java:2681)
	at net.minecraft.client.Minecraft.main(Minecraft.java:2787)
	at net.fabricmc.loader.impl.game.minecraft.MinecraftGameProvider.launch(MinecraftGameProvider.java:475)
	at net.fabricmc.loader.impl.launch.knot.Knot.launch(Knot.java:74)
	at net.fabricmc.loader.impl.launch.knot.KnotClient.main(KnotClient.java:23)
	at net.fabricmc.devlaunchinjector.Main.main(Main.java:86)




```

After these issues are fixed, I'd be ready to merge and release :D

IsLocal committed:
Updated loom version, added lwjgl dependencies, resolved some issues.

IsLocal commented:
Yeah seems the build.gradle needed to be updated. Only reason it was compiling for me was I had all the dependencies in my cache.

vinceingithub commented:
Will this ever be merged and a new build released? I want to use this.

5Avian commented:
Perhaps some day. I likely won't be spending any time on programming related things in the near future, so if you wanna use this version, I recommend building it yourself. The MIT license that this project is under allows you to do basically anything with the project, including (forking it and) releasing it under your name on platforms like CurseForge and Modrinth, as long as you include the exact license text for this project in source code and releases.

IsLocal commented:
> Will this ever be merged and a new build released? I want to use this.

There’s a release of this under my fork of the project.
